src/main/java/core/basesyntax/impl/StorageImpl.java

```diff
@@ -1,19 +1,144 @@
 package core.basesyntax.impl;
 
 import core.basesyntax.Storage;
+import java.lang.reflect.Array;
 
 public class StorageImpl<K, V> implements Storage<K, V> {
+    private static final int MAX_LENGTH = 10;
+    private int currentLength;
+    private K[] keys;
+    private V[] values;
+
+    public StorageImpl() {
+        currentLength = 0;
+    }
+
     @Override
     public void put(K key, V value) {
+        initializeArraysIfNeeded(key, value);
+        if (addedValueNoNullArray(value)) {
+            return;
+        }
+        if (replacedValueByExistingKey(key, value)) {
+            return;
+        }
+        if (currentLength == MAX_LENGTH) {
+            throw new RuntimeException("Storage is full, can't add any more objects");
+        }
+        addNewRowToStorage(key, value);
+    }
+
+    private void initializeArraysIfNeeded(K key, V value) {
+        if (keys == null && key != null) {
+            keys = (K[]) Array.newInstance(key.getClass(), MAX_LENGTH);
+        }
+        if (values == null && value != null) {
+            values = (V[]) Array.newInstance(value.getClass(), MAX_LENGTH);
+        }
+    }
+
+    private boolean addedValueNoNullArray(V value) {
+        //if keys aren't initialized yet but we need to add a value,
+        //we are writing or rewriting value to [0] cell
+        if (keys == null && values != null) {
+            values[0] = value;
+            if (currentLength == 0) {
+                currentLength++;
+            }
+            return true;
+        }
+        return false;
+    }
+
+    private boolean replacedValueByExistingKey(K key, V value) {
+        for (int i = 0; i < currentLength; i++) {
+            if (!(keys[i] == null && key != null)
+                    && ((keys[i] == null && key == null) || keys[i].equals(key))) {
+                values[i] = value;
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private void addNewRowToStorage(K key, V value) {
+        if (keys != null) {
+            keys[currentLength] = key;
+        }
+        if (values != null) {
+            values[currentLength] = value;
+        }
+        currentLength++;
     }
 
     @Override
     public V get(K key) {
+        if (keys == null) {
+            return values == null ? null : values[0];
+        }
+        for (int i = 0; i < currentLength; i++) {
+            if ((keys[i] == null) == (key == null)
```

Comment:
```suggestion
            if ((keys[i] == key)
```
якщо два об'єкта нал - порівняння по посиланню відпрацюють коректно, якщо ні у насє друга частина іфу

Reply:
Если сделать как ты указал, получаю ошибку тестов.

У меня ведь целиком условие так выглядит:
if ((keys[i] == null) == (key == null)
                    && (key == null || keys[i].equals(key))) {

И если в первой половине я поставлю "(keys[i] == key)", то это значит, что ключи должны совпадать по ссылке, а тогда валится тест "getElementWhenKeyIsObject", где сравнивают "elementTwo" и storage.get(sameSecondCat)".

Если же твоё "if ((keys[i] == key)" заменить на "if ((keys[i].equals(key))", то валится тест "getElementWithKeyNull", так как при key = null срабатывает ошибка "java.lang.NullPointerException: Cannot invoke "Object.equals(Object)" because "this.keys[i]" is null"